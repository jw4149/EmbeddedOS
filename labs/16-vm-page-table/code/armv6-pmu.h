// single header file code to setup/use the armv6 performance
// monitor unit (PMU) -- see 3-134 arm1176.pdf
//
// we abuse the hell out of the preprocessor using an old trick
// from the lcc compiler to generate all the helpers and
// enums.
//
// NOTE: the r/pi does not propagate the PMU interrupt, 
// unfortunately, so there is no way to do a sampling 
// profiler by using it.
#ifndef __ARMV6_PMU_H__
#define __ARMV6_PMU_H__

#include "bit-support.h"
#include "asm-helpers.h"

// use the <cp_asm> macro to define the different 
// get/set PMU instructions.
cp_asm(pmu_cycle, p15, 0, c15, c12, 1)
cp_asm(pmu_event0, p15, 0, c15, c12, 2)
cp_asm(pmu_event1, p15, 0, c15, c12, 3)
cp_asm(pmu_control, p15, 0, c15, c12, 0)

// p 3-134 arm1176.pdf: configure the PMU control register
static inline void pmu_control_config(uint32_t in) {
    // 3-134 says to clear these.

    // you have to *set* these bits to clear the event.
    in = bits_set(in, 8, 10, 0b111);
    assert(bits_get(in, 8,10) == 0b111);

    pmu_control_set(in);
}

// get the type of event0 by reading the type
// field from the PMU control register and 
// returning it.
static inline uint32_t pmu_type0(void) {
    uint32_t r = pmu_control_get();
    return bits_get(r, 20, 27);
}

// set PMU event0 as <type> and enable it.
static inline void pmu_enable0(uint32_t type) {
    assert((type & 0xff) == type);

    uint32_t r = pmu_control_get();
    r = bits_set(r, 20, 27, type);
    pmu_control_config(r | 0b11);

    assert(pmu_type0() == type);
}

// get the type of event1 by reading the type
// field from the PMU control register and 
// returning it.
static inline uint32_t pmu_type1(void) {
    uint32_t r = pmu_control_get();
    return bits_get(r, 12, 19);
}

// set event1 as <type> and enable it.
static inline void pmu_enable1(uint32_t type) {
    assert((type & 0xff) == type);

    uint32_t r = pmu_control_get();
    r = bits_set(r, 12, 19, type);
    pmu_control_config(r | 0b11);

    assert(pmu_type1() == type);
}

// wrapper so can pass in the PMU register number.
static inline void pmu_enable(unsigned n, uint32_t type) {
    if(n==0)
        pmu_enable0(type);
    else if(n == 1)
        pmu_enable1(type);
    else
        panic("bad PMU coprocessor number=%d\n", n);
}

// wrapper so can pass in the PMU register number.
static inline uint32_t pmu_event_get(unsigned n) {
    if(n==0)
        return pmu_event0_get();
    else if(n == 1)
        return pmu_event1_get();
    else
        panic("bad PMU coprocessor number=%d\n", n);
}

// measure and print the count of <type0> and <type1>
// events when running the statements <stmts>.
//   - will write the results in <cnt0> and <cnt1>
//
// we could make it so it records the current
// and then restores it.  not sure if this 
// makes any sense.
#define pmu_stmt_measure_set(cnt0, cnt1, msg, type0, type1, stmts) do { \
    pmu_ ## type0 ## _on(0);                            \
    pmu_ ## type1 ## _on(1);                            \
    uint32_t ty0 = pmu_ ## type0(0);                    \
    uint32_t ty1 = pmu_ ## type1(1);                    \
    uint32_t cyc = pmu_cycle_get();                     \
                                                        \
    /* partially stop gcc from moving stuff */          \
    gcc_mb();                                           \
    stmts;                                              \
    gcc_mb();                                           \
                                                        \
    uint32_t n_ty0 = pmu_ ## type0(0) - ty0;            \
    uint32_t n_ty1 = pmu_ ## type1(1) - ty1;            \
    uint32_t n_cyc = pmu_cycle_get() -  cyc;            \
    const char *s0 = pmu_ ## type0 ## _str();           \
    const char *s1 = pmu_ ## type1 ## _str();           \
                                                        \
    cnt0 = n_ty0;                                       \
    cnt1 = n_ty1;                                       \
                                                        \
    output("%s:%d: %s:\n\tcycles=%d\n\t%s=%d\n\t%s=%d\n", \
        __FILE__, __LINE__, msg, n_cyc, s0, cnt0, s1, cnt1);              \
} while(0)

#define pmu_stmt_measure(msg, type0, type1, stmts) do {     \
    uint32_t c0 __attribute__((unused));                    \
    uint32_t c1 __attribute__((unused));                    \
    pmu_stmt_measure_set(c0,c1,msg, type0, type1, stmts);   \
} while(0)

// define the different event enums values.
enum {
#   define XX(fn,name,val,string) name = val,
#   define PMU_GET_TYPES

#   include "armv6-pmu.h"

#   undef PMU_GET_TYPES
#   undef XX
};

// generate the different event wrappers for setting
// reading.
#   define XX(fn_name,type,val,string)                      \
    /* enable event <type> */                               \
    static void pmu_ ## fn_name ## _on(unsigned n)          \
        { pmu_enable(n, type); }                            \
    /* read event <type> counter */                         \
    static uint32_t pmu_ ## fn_name(unsigned n)             \
        { return pmu_event_get(n); }                        \
    /* get printable string of event <type>*/               \
    static const char * pmu_ ## fn_name ## _str(void)       \
        { return string; }

#   define PMU_GET_TYPES
#   include "armv6-pmu.h"
#   undef XX
#   undef PMU_GET_TYPES

#endif

#ifdef PMU_GET_TYPES
// see 3-139 in arm1176.pdf
XX(ret_miss,        PMU_RET_MISPREDICTED,    0x26, "return mispredicted")
XX(ret_hit,         PMU_RET_PREDICTED,       0x25, "return predicted")
XX(ret_cnt,         PMU_RET_CNT,             0x24, "return count")
// the addresss was pushed on the return stack.
// there is also a 0x24 where it is popped off the 
// stack.  idk difference.
XX(call_cnt,        PMU_CALL_CNT,            0x23, "call count")
XX(main_tlb_miss,   PMU_TLB_MISS,            0x16, "TLB miss")
XX(wb_drain, PMU_WBACK_DRAINED,       0x12, "writeback drained (stall?)")
// data cache miss, noncacheable, write through.
XX(data_access,     PMU_EXPLICIT_DATA_ACCESS,0x10, "explicit data access")

XX(tlb_miss,        PMU_MAIN_TLB_MISS,       0xF,  "main tlb miss")
// i think this is all indirect jumps
XX(pc_write,        PMU_PC_WRITE,            0xD,  "sw changed pc [indir jump?]")
XX(dcache_wb,       PMU_DCACHE_WB,           0xC,  "dcache write back")
XX(dcache_miss,     PMU_DCACHE_MISS,         0xB,  "dcache miss")
XX(dcache_access,   PMU_DCACHE_ACCESS,       0xA,  "dcache access")
XX(inst_cnt,        PMU_INST_CNT,            0x7,  "instruction count")
XX(branch_miss,     PMU_BRANCH_MISPREDICT,   0x6,  "branch mispredict")
XX(branch_cnt,      PMU_BRANCH_EXECUTED,     0x5,  "branch executed")
XX(dtlb_miss, PMU_DATA_MICROTLB_MISS,  0x4,  "data microtlb miss")
XX(itlb_miss, PMU_INST_MICROTLB_MISS,  0x3,  "instruction microtlb miss")
XX(data_stall,      PMU_DATA_STALL,          0x2,  "data dependency stall")
XX(inst_stall,      PMU_INST_STALL,          0x1,  "instruction stall [tlb/cache miss]")
XX(icache_miss,     PMU_ICACHE_MISS,         0x0,  "icache miss")
XX(cycle_cnt,       PMU_CYCLE_CNT,           0xff, "cycle count")
#endif
