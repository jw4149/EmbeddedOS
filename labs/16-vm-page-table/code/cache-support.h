#ifndef __CACHE_SUPPORT_H__
#define __CACHE_SUPPORT_H__
// turn on the different caches by changing 
// co-processor 15.  
//
// note: unclear if the r/pi lets the CPU use the
// L2 cache directly.  if it does there is some
// additional configuration we need to do.
#include "armv6-cp15.h"

// libpi/staff-src/cache-support.S
// flushes tlb, icache dcache etc.
void cache_flush_all(void);

// you have to define this.
int mmu_is_enabled(void);

// enable btc
static inline int btc_is_on(void) {
    let c = cp15_ctrl_reg1_rd();
    return c.Z_branch_pred == 1;
}

// branch target cache.
static inline void btc_on(void) {
    assert(mmu_is_enabled());
    let c = cp15_ctrl_reg1_rd();
    c.Z_branch_pred = 1;
    cp15_ctrl_reg1_wr(c);
    assert(btc_is_on());
}

static inline int dcache_l1_is_on(void) {
    let c = cp15_ctrl_reg1_rd();
    return c.C_unified_enable == 1;
}

// enable just the L1 cache: doesn't work
// without mmu on.
static inline void dcache_l1_on(void) {
    assert(mmu_is_enabled());
    let c = cp15_ctrl_reg1_rd();
    // assert(c.C_unified_enable == 0);
    c.C_unified_enable = 1;
    cp15_ctrl_reg1_wr(c);
    assert(dcache_l1_is_on());
}


// enable dcache writeback: doesn't work 
// without mmu on.
static inline int dcache_wb_is_on(void) {
    let c = cp15_ctrl_reg1_rd();
    return c.W_write_buf == 1;
}

static inline void dcache_wb_on(void) {
    assert(mmu_is_enabled());
    let c = cp15_ctrl_reg1_rd();
    c.W_write_buf = 1;
    cp15_ctrl_reg1_wr(c);

    assert(dcache_wb_is_on());
}


static inline int dcache_l2_is_on(void) {
    let c = cp15_ctrl_reg1_rd();
    return c.L2_enabled == 1;
}

// enable l2 cache.
static inline void dcache_l2_on(void) {
    panic("need to do extra stuff to enable\n");
    assert(mmu_is_enabled());
    let c = cp15_ctrl_reg1_rd();
    c.L2_enabled = 1;
    cp15_ctrl_reg1_wr(c);
    assert(dcache_l2_is_on());
}

// disable all caches: note, you better make
// sure you flush and invalidate everything
// first.
static inline void caches_all_off(void) {
    // i think you need to flush it first.  interesting
    // bug.
    cache_flush_all();

    let c = cp15_ctrl_reg1_rd();
    c.C_unified_enable = 0;
    c.W_write_buf = 0;
    c.L2_enabled = 0;
    c.Z_branch_pred = 0;
    c.I_icache_enable = 0;
    cp15_ctrl_reg1_wr(c);
}

static inline int icache_is_on(void) {
    let c = cp15_ctrl_reg1_rd();
    return c.I_icache_enable == 1;
}
static inline void icache_on(void) {
    let c = cp15_ctrl_reg1_rd();
    c.I_icache_enable = 1;
    cp15_ctrl_reg1_wr(c);

    assert(icache_is_on());
}

static inline int caches_all_on_p(void) {
    return 
        btc_is_on()
        && dcache_wb_is_on()
        && dcache_l1_is_on()
        && icache_is_on();
}

// enable all caches.
static inline void caches_all_on(void) {
    assert(mmu_is_enabled());

    icache_on();
    btc_on();
    dcache_l1_on();
    dcache_wb_on();
    // dcache_l2_on();

#if 0
    // faster way but we do the slow way
    // for internal consistency checking
    let c = cp15_ctrl_reg1_rd();
    c.C_unified_enable = 1;
    c.W_write_buf = 1;
    c.L2_enabled = 1;
    c.Z_branch_pred = 1;
    c.I_icache_enable = 1;
    cp15_ctrl_reg1_wr(c);
#endif
    
    assert(caches_all_on_p());
}

#endif
