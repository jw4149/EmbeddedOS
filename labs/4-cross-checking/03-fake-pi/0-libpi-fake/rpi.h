#ifndef __RPI_H__
#define __RPI_H__
// simplest way to rewrap --- just define replacements in header --- don't
// even need .c
//
// easy, but two downsides: 
//   1. we are polluting the pi program namespace with system includes.
//   2. we aren't using the raw rpi.h so can get different behavior.
#include <stdio.h>
#include <stdlib.h>

// redefine printk as fprintf using variadic macros.
#define printk(args...) fprintf(stderr, args)

static void inline clean_reboot(void) {
    printk("DONE!!!\n");
    exit(0);
}

void notmain(void);
#endif
