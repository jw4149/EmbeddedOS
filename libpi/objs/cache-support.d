objs/cache-support.o: staff-src/cache-support.S \
 /Users/jiayangwang/Desktop/cs140e-24win//libpi/include/rpi-asm.h \
 /Users/jiayangwang/Desktop/cs140e-24win//libpi/include/rpi-constants.h

/Users/jiayangwang/Desktop/cs140e-24win//libpi/include/rpi-asm.h:

/Users/jiayangwang/Desktop/cs140e-24win//libpi/include/rpi-constants.h:
