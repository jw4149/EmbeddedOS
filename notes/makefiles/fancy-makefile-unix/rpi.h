#include <stdio.h>

#define printk printf
#define debug printf
