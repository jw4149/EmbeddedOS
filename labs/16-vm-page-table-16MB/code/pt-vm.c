#include "rpi.h"
#include "pt-vm.h"
#include "helper-macros.h"
#include "procmap.h"

// turn this off if you don't want all the debug output.
enum { verbose_p = 1 };
enum { OneMB = 1024*1024 };

vm_pt_t *vm_pt_alloc(unsigned n) {
    demand(n == 4096, we only handling a fully-populated page table right now);

    vm_pt_t *pt = 0;
    unsigned nbytes = 4096 * sizeof *pt;

    // allocate pt with n entries.
    // pt = staff_vm_pt_alloc(n);
    pt = kmalloc(nbytes);
    memset(pt, 0, nbytes);
    demand(is_aligned_ptr(pt, 1<<14), must be 14-bit aligned!);
    return pt;
}

// allocate new page table and copy pt.
vm_pt_t *vm_dup(vm_pt_t *pt1) {
    vm_pt_t *pt2 = vm_pt_alloc(PT_LEVEL1_N);
    memcpy(pt2,pt1,PT_LEVEL1_N * sizeof *pt1);
    return pt2;
}

// same as pinned version: probably should check that
// the page table is set.
void vm_mmu_enable(void) {
    assert(!mmu_is_enabled());
    mmu_enable();
    assert(mmu_is_enabled());
}

// same as pinned version.
void vm_mmu_disable(void) {
    assert(mmu_is_enabled());
    mmu_disable();
    assert(!mmu_is_enabled());
}

// - set <pt,pid,asid> for an address space.
// - must be done before you switch into it!
// - mmu can be off or on.
void vm_mmu_switch(vm_pt_t *pt, uint32_t pid, uint32_t asid) {
    assert(pt);
    mmu_set_ctx(pid, asid, pt);
}

// just like pinned.
void vm_mmu_init(uint32_t domain_reg) {
    // initialize everything, after bootup.
    mmu_init();
    domain_access_ctrl_set(domain_reg);
}

// map the 1mb section starting at <va> to <pa>
// with memory attribute <attr>.
vm_pte_t *
vm_map_sec(vm_pt_t *pt, uint32_t va, uint32_t pa, pin_t attr) 
{
    assert(aligned(va, OneMB));
    assert(aligned(pa, OneMB));

    // today we just use 1mb.
    assert(attr.pagesize == PAGE_1MB);

    unsigned index = va >> 20;
    assert(index < PT_LEVEL1_N);

    vm_pte_t *pte = &pt[index];
    // return staff_vm_map_sec(pt,va,pa,attr);
    pte->tag = 0b10;
    pte->B = attr.mem_attr & 0x1;
    pte->C = (attr.mem_attr >> 1) & 0x1;
    pte->XN = 0b0;
    pte->domain = attr.dom;
    pte->IMP = 0b0;
    pte->AP = attr.AP_perm & 0x11;
    pte->TEX = attr.mem_attr >> 2;
    pte->APX = attr.AP_perm >> 2;
    pte->S = 0b0;
    pte->nG = ~attr.G;
    pte->super = 0b0;
    pte->sec_base_addr = pa >> 20;

    if(verbose_p)
        vm_pte_print(pt,pte);
    assert(pte);
    return pte;
}

// lookup 32-bit address va in pt and return the pte
// if it exists, 0 otherwise.
vm_pte_t * vm_lookup(vm_pt_t *pt, uint32_t va) {
    // return staff_vm_lookup(pt,va);
    uint32_t index = va >> 20;
    vm_pte_t * pte = &pt[index];
    if (pte->tag != 0b10) {
        return 0;
    }
    return pte;
}

// manually translate <va> in page table <pt>
// - if doesn't exist, returns 0.
// - if does exist returns the corresponding physical
//   address in <pa>
//
// NOTE: we can't just return the result b/c page 0 could be mapped.
vm_pte_t *vm_xlate(uint32_t *pa, vm_pt_t *pt, uint32_t va) {
    // return staff_vm_xlate(pa,pt,va);
    vm_pte_t * pte = vm_lookup(pt, va);
    if (!pte) {
        return 0;
    }
    *pa = (pte->sec_base_addr) << 20;
    return pte;
}

// compute the default attribute for each type.
static inline pin_t attr_mk(pr_ent_t *e) {
    switch(e->type) {
    case MEM_DEVICE: 
        return pin_mk_device(e->dom);
    // kernel: currently everything is uncached.
    case MEM_RW:
        return pin_mk_global(e->dom, perm_rw_priv, MEM_uncached);
   case MEM_RO: 
        panic("not handling\n");
   default: 
        panic("unknown type: %d\n", e->type);
    }
}

// setup the initial kernel mapping.  This will mirror
//  static inline void procmap_pin_on(procmap_t *p) 
// in <15-pinned-vm/code/procmap.h>  but will call
// your vm_ routines, not pinned routines.
//
// if <enable_p>=1, should enable the MMU.  make sure
// you setup the page table and asid. use  
// kern_asid, and kern_pid.
vm_pt_t *vm_map_kernel(procmap_t *p, int enable_p) {
    // the asid and pid we start with.  
    //    shouldn't matter since kernel is global.
    enum { kern_asid = 1, kern_pid = 0x140e };

    vm_pt_t *pt = 0;

    // return staff_vm_map_kernel(p,enable_p);
    uint32_t d = dom_perm(p->dom_ids, DOM_client);
    vm_mmu_init(d);
    
    pt = vm_pt_alloc(4096);

    for (int i = 0; i < p->n; i++) {
        pr_ent_t pr_ent = p->map[i];
        pin_t attr = attr_mk(&pr_ent);
        vm_map_sec(pt, pr_ent.addr, pr_ent.addr, attr);
    }

    vm_mmu_switch(pt, kern_pid, kern_asid);

    if (enable_p) {
        vm_mmu_enable();
    }

    assert(pt);
    return pt;
}
