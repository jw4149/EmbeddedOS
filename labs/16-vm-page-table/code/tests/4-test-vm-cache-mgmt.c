// a very incomplete set of tests that use the arm PMU counters
// to check that different VM operations work correctly.

#include "rpi.h"
#include "pt-vm.h"
#include "armv6-pmu.h"
#include "cache-support.h"

enum { cache_addr = MB * 10 };

// load a single pointer: should see one dcache miss if 
// not in cache.
static void
dcache_should_miss(const char *msg, volatile uint32_t *ptr) {
    uint32_t miss, cnt;

    // compulsory miss: we've never used <*ptr>
    pmu_stmt_measure_set(miss,cnt,
        msg,
        dcache_miss, inst_cnt,
        { 
            *ptr; 
        });

    if(miss == 0)
        panic("should have at least 1 miss, have 0\n");
}

// do 5 loads: should see zero dcache misses if in cache.
static void 
dcache_should_not_miss(const char *msg, volatile uint32_t *ptr) {
    uint32_t miss, access;

    // should not miss.
    pmu_stmt_measure_set(miss, access, 
        "should be cached: should see no misses", 
        dcache_miss, dcache_access,
        { 
            *ptr;  /* 1 */
            *ptr;  /* 2 */
            *ptr;  /* 3 */
            *ptr;  /* 4 */
            *ptr;  /* 5 */
        });
    if(miss > 0)
        panic("should have had zero misses: have %d\n", miss);
    if(access <  5)
        panic("expected 5 accesses have %d\n", access);
}

// call fp [128 nops + 1 bx lr]
// if not in icache should be 32 prefetch buffer fetches.
static void 
icache_should_miss(const char *msg, uint32_t (*fp)(uint32_t)) {
    uint32_t miss,cnt;

    // checking icache is hard we can get conflicts or pull other
    // stuff in.
    //
    // weird: it appears that only the initial prefetch buffer fetch 
    // counts as a icache miss?
    pmu_stmt_measure_set(miss,cnt,
        msg,
        icache_miss, inst_cnt,
        { 
            fp(1);
        });

    // we should have a ton of misses on fp: 
    //  128 instructions / 4 inst prefetch buff = 32 misses.
    if(miss < 32)
        panic("should have at least 32 prefetch buffer misses\n");
}

// helper to do a raw dache invalidation: if there are dirty
// blocks they will be loss (you can see b/c subsequent reads
// will return old copies)
static inline void dcache_inv(void) {
    uint32_t r=0;
    asm volatile ("mcr p15, 0, %0, c7, c6, 0" :: "r"(r));
    prefetch_flush();
}

void notmain(void) { 
    // 
    // sleazy hack: create a routine at runtime that has
    // 256 nops and one bx lr on the cached page.
    //
    // we do this before vm/dcache/icache on.
    volatile uint32_t *code_ptr = (void*)(cache_addr+128);
    unsigned i;
    for(i = 0; i < 256; i++)
        code_ptr[i] = 0xe320f000; // nop
    code_ptr[i] = 0xe12fff1e;   // bx lr
    dmb();
    uint32_t (*fp)(uint32_t x) = (void *)code_ptr;
    assert(fp(5) == 5);
    assert(fp(2) == 2);


    // map the heap: for lab cksums must be at 0x100000.
    kmalloc_init_set_start((void*)MB, MB);
    enum { dom_kern = 1 };

    procmap_t p = procmap_default_mk(dom_kern);
    vm_pt_t *pt = vm_map_kernel(&p,0);
    assert(!mmu_is_enabled());

    // map a single cached page with writeback allocation.
    // write back, write alloc
    enum { wb_alloc     =  TEX_C_B(    0b001,  1, 1) };

    pin_t attr = pin_mk_global(dom_kern, perm_rw_priv, wb_alloc);

    vm_map_sec(pt, cache_addr, cache_addr, attr);
    vm_mmu_enable();
    assert(mmu_is_enabled());

    volatile uint32_t *ptr = (void*)cache_addr;

    // turn all caches on.
    caches_all_on();
    assert(caches_all_on_p());

    uint32_t miss,access;

    trace("******************************************************\n");
    trace("1. test that the page table enabled dcaching correctly.\n");
    // compulsory: tests the counters.
    dcache_should_miss("first access: should see one miss", ptr);

    // now <ptr> should be in the cache.
    dcache_should_not_miss("should be cached", ptr);

    trace("******************************************************\n");
    trace("2. test that the page table enabled icaching correctly.\n");

    // get the baseline misses i think it's zero b/c of prefetch buff.
    uint32_t base_miss = 0, cnt;
    pmu_stmt_measure_set(base_miss, cnt,
        "testing miss on nop",
        icache_miss, inst_cnt,
        {
            asm volatile ("nop");
        });

    icache_should_miss("first access: should see many icache misses", fp);
    assert(caches_all_on_p());

    trace("******************************************************\n");
    trace("3. test that sync pte mods flushes the dcache.\n");
    // sync pte mods better flush the dcache.
    mmu_sync_pte_mods();
    dcache_should_miss("after pte mod: dcache should be flushed", ptr);
    icache_should_miss("after pte mod: icache should be flushed", fp);

    trace("******************************************************\n");
    trace("4. test that mmu_on/off flushes the dcache.\n");
    assert(caches_all_on_p());
    vm_mmu_disable();
    vm_mmu_enable();
    assert(caches_all_on_p());

    // mmu_disable should have cleaned/invalidated dcache and icache.
    dcache_should_miss("after mmu on/off: dcache should be flushed", ptr);
    icache_should_miss("after mmu on/off: icache should be flushed", fp);

    trace("******************************************************\n");
    trace("5. about to test that disable dcache invalidates correctly\n");

    // check if we lose writes to cached memory.
    for(unsigned i =0; i < 512; i++)  {
        *ptr = i;
        vm_mmu_disable();
        vm_mmu_enable();
        assert(caches_all_on_p());

        if(*ptr != i)
            panic("lost a write!  have %d, expected=%d!\n", *ptr,i);
    }

    trace("******************************************************\n");
    trace("6. about to test sync_pte_mods invalides dcache correctly\n");

    // check if we lose writes to cached memory.
    for(unsigned i =0; i < 512; i++)  {
        *ptr = i;
        mmu_sync_pte_mods();
        assert(caches_all_on_p());

        if(*ptr != i)
            panic("lost a write!  have %d, expected=%d!\n", *ptr,i);
    }


    trace("SUCCESS!\n");
}
