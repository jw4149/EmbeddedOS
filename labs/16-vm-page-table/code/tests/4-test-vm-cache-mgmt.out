TRACE:notmain:******************************************************
TRACE:notmain:1. test that the page table enabled dcaching correctly.
TRACE:notmain:******************************************************
TRACE:notmain:2. test that the page table enabled icaching correctly.
TRACE:notmain:******************************************************
TRACE:notmain:3. test that sync pte mods flushes the dcache.
TRACE:notmain:******************************************************
TRACE:notmain:4. test that mmu_on/off flushes the dcache.
TRACE:notmain:******************************************************
TRACE:notmain:5. about to test that disable dcache invalidates correctly
TRACE:notmain:******************************************************
TRACE:notmain:6. about to test sync_pte_mods invalides dcache correctly
TRACE:notmain:SUCCESS!
